### ISSUE TITLE
Add manifold version of LargeVis

### ISSUE THREAD

**Issue by josevalim** [elixir-nx/scholar#272]
We already have the building blocks for ANN. @krstopro, how much work do you believe it is to also have it for manifold/dimensionality reduction?

**Comment by krstopro** [comment]
Probably not much, though I am worried about memory cost of current implementations. `LargeVis` dimensionality reduction uses large `num_neighbors` (150 in the paper) and `RandomProjectionForest.predict/2` takes $O(QTK)$ memory for number of queries $Q$, number of trees $T$, and number of neighbors $K$. In any case, leave it to me.

By the way, thanks for the shout out on the blog. :)

**Comment by krstopro** [comment]
Apparently, this is much harder than I thought. One of the issues is sampling the negative edges of the k-NN graph, i.e. edges that are not in the graph. To do this, [the authors allocate and iterate over a rather large table](https://github.com/lferry007/LargeVis/blob/feb8121e8eb9652477f7f564903d189ee663796f/Linux/LargeVis.cpp#L544-L567). I don't like this approach and I don't think I can come up with something efficiently implementable using `Nx`.

**Comment by josevalim** [comment]
Why do you think it would be hard to implement it? The parr that seems complex is the head traversal, but if we can assume a max depth (not sure if that’s the case), it should be parallelizable.

**Comment by krstopro** [comment]
It's not really about head traversal, more about allocating a table of size `1e8`. I'm still checking if this can be avoided.

**Comment by krstopro** [comment]
The problem above might be avoidable, but there is another one that might not be. Namely, the method involves symmetrising the k-NN graph and this requires the computation of reverse k-NN graph which I don't think can be done efficiently with `Nx`.
<img width="468" alt="Screenshot 2024-06-14 at 19 03 13" src="https://github.com/elixir-nx/scholar/assets/7234787/9b7a8988-97e9-41a6-be00-34500617eb1d">

**Comment by josevalim** [comment]
The symmetrized seems to be doable with a while loop over i,j?

**Comment by krstopro** [comment]
Even if it is possible, it would be expensive as we would need to loop over $n \times k$ edges. A bigger issue is how to store the weights. $p_{j | i}$ is $n \times k$ matrix (or sparse $n \times n$ matrix) since we take $k$-nearest neighbors for each of $n$ points; each  of $n$ nodes in $k$-NN graph has a degree of $k$. $p_{i | j}$ requires reversing the $k$-NN graph and the degree of every node can vary from $0$ (the point is not among $k$-nearest neighbors of any other point) to $n - 1$ (the point is contained among $k$-nearest neighbors of every other point). I cannot think of a way to represent the weights with a tensor.

**Comment by RicardoSantos-99** [comment]
Hi @krstopro, I saw this has been open for a while and wanted to check before touching it, since you're assigned. Are you still actively working on this, or would you be open to me giving it a shot?

I looked into the reverse-graph blocker you ran into and think there's a way around it: treating the k-NN graph as a fixed-size edge list (source, target, weight) instead of an explicit graph structure. The "reverse" is just that list with the columns swapped, still fixed size, so the variable in-degree never needs to be represented. I've put together a rough skeleton based on this, still early and untested in a lot of ways, but wanted to check in before going further.

**Comment by krstopro** [comment]
> I looked into the reverse-graph blocker you ran into and think there's a way around it: treating the k-NN graph as a fixed-size edge list (source, target, weight) instead of an explicit graph structure. The "reverse" is just that list with the columns swapped, still fixed size, so the variable in-degree never needs to be represented. I've put together a rough skeleton based on this, still early and untested in a lot of ways, but wanted to check in before going further.

The problem is the following. $k$-NN graph can be represented as $n \times k$ matrix simply by storing the indices of $k$-nearest neighbors of all the $n$ points. You cannot do the same with the reverse graph; it could be the case that one of the points is among $k$-nearest neighbors of every other point. Therefore, the reverse graph cannot be represented using a fixed-size structure; its degree depends on the $k$-NN graph itself. Does this make sense?

**Comment by RicardoSantos-99** [comment]
> > I looked into the reverse-graph blocker you ran into and think there's a way around it: treating the k-NN graph as a fixed-size edge list (source, target, weight) instead of an explicit graph structure. The "reverse" is just that list with the columns swapped, still fixed size, so the variable in-degree never needs to be represented. I've put together a rough skeleton based on this, still early and untested in a lot of ways, but wanted to check in before going further.
> 
> The problem is the following. k -NN graph can be represented as n × k matrix simply by storing the indices of k -nearest neighbors of all the n points. You cannot do the same with the reverse graph; it could be the case that one of the points is among k -nearest neighbors of every other point. Therefore, the reverse graph cannot be represented using a fixed-size structure; its degree depends on the k -NN graph itself. Does this make sense?

I think it's answering a slightly different question than what I had in mind. Let me try to make it concrete with a tiny example.

Say n=4, k=2, and the k-NN graph looks like:

```
point 0 -> neighbors [1, 2]
point 1 -> neighbors [0, 2]
point 2 -> neighbors [0, 1]
point 3 -> neighbors [0, 1]
```

Point 3's reverse degree is 0, nobody has it as a neighbor. Points 0 and 1's reverse degree is 3, everybody else points at them. That's basically the exact case you're worried about, and you're right, that can't be grouped into a fixed-shape "list of reverse neighbors per node."

But that's not quite what I'm doing. Instead of grouping by node at all, just treat the whole thing as a flat list of individual edges, one row per (source, target, weight), forward and reverse both:

```
forward edges (n*k = 8, always):
(0,1) (0,2) (1,0) (1,2) (2,0) (2,1) (3,0) (3,1)

reverse = same list, columns swapped (still 8):
(1,0) (2,0) (0,1) (2,1) (0,2) (1,2) (0,3) (1,3)
```

Concatenated that's 16 rows, always 2nk no matter how lopsided the degrees get. Point 3 going from in-degree 0 to whatever didn't change that count one bit. If you scatter-add all 16 triples into a dense matrix you get exactly P + P^T: (0,3) picks up one hit (from the reverse-swapped (3,0) becoming (0,3)), (0,1) picks up two (one from each direction), matches the dense version exactly.

The reason it works out is the optimizer afterwards never asks "give me all of node i's neighbors," which is where the variable degree would actually bite you. It only ever asks "give me a random edge, weighted by probability" and nudges just those two points. So the per-node grouping, which you're right to say doesn't fit in a tensor, never actually gets asked for anywhere in the pipeline.

**Comment by krstopro** [comment]
> But that's not quite what I'm doing. Instead of grouping by node at all, just treat the whole thing as a flat list of individual edges, one row per (source, target, weight), forward and reverse both:

What you are doing is building a sparse matrix and transposing it. Reversing the graph is efficient in this approach. Some other operations might not be, e.g. symmetrising the graph as mentioned above.

**Comment by RicardoSantos-99** [comment]
> What you are doing is building a sparse matrix and transposing it. Reversing the graph is efficient in this approach. Some other operations might not be, e.g. symmetrising the graph as mentioned above.

Good point. I don't actually deduplicate though, the SGD sampling handles the combined weight implicitly, so that part stays cheap too. Let me put together something more concrete along these lines and share it so you can see if it holds up.

**Comment by RicardoSantos-99** [comment]
@krstopro 

The concrete version is up in #345 

On your point about symmetrising being the expensive part: it never symmetrises in the sense of forming P + P^T. The k-NN graph stays as a flat edge list of (source, target, weight) triples of size nk. The reverse is that same list with the two columns swapped, and the two are just concatenated into one 2n*k list. That concatenation is all the "symmetrisation" there is, no scatter into an n by n matrix and no grouping by node anywhere.

Why that is enough: the only thing the optimiser asks of the graph is "give me a random edge, weighted by probability", which Nx.Random.choice does over the flat weight vector. It never asks "give me all edges incident to node i", which is the query that would need the variable in-degree. Duplicate (i, j) rows are fine, they just get sampled proportionally more often, which is exactly what summing their weights would give.

I also checked the identity numerically: scatter-adding the 2nk edge list into a dense matrix reproduces P + P^T exactly (max diff 0.0), including the degenerate case where one node is every other point's neighbour. [build_edges](https://github.com/elixir-nx/scholar/pull/345/files#diff-244737a33104c0334327b04711dfc37dbfda693c5451cc3f87510417e4e9ff62R271-R281) is where this happens if you want to trace it.
